src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/UseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,193 @@
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseParameterNullChecking
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class UseParameterNullCheckingDiagnosticAnalyzer
+        : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private const string ArgumentNullExceptionName = $"{nameof(System)}.{nameof(ArgumentNullException)}";
+
+        public UseParameterNullCheckingDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UseParameterNullCheckingId,
+                   EnforceOnBuildValues.UseParameterNullChecking,
+                   CodeStyleOptions2.PreferParameterNullChecking,
+                   CSharpAnalyzersResources.Use_parameter_null_checking,
+                   new LocalizableResourceString(nameof(AnalyzersResources.Null_check_can_be_simplified), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(context =>
+            {
+                if (((CSharpCompilation)context.Compilation).LanguageVersion < LanguageVersionExtensions.CSharpNext)
+                {
+                    return;
+                }
+
+                var compilation = context.Compilation;
+                var argumentNullException = compilation.GetTypeByMetadataName(ArgumentNullExceptionName);
+                if (argumentNullException is null)
+                {
+                    return;
+                }
+
+                var objectType = compilation.GetSpecialType(SpecialType.System_Object);
+                var referenceEqualsMethod = objectType
+                    .GetMembers(nameof(ReferenceEquals))
+                    .OfType<IMethodSymbol>()
+                    .FirstOrDefault(m => m.DeclaredAccessibility == Accessibility.Public && m.Parameters.Length == 2);
+                // This seems sufficiently rare that it's not a big deal to just not perform analysis in this case.
+                if (referenceEqualsMethod is null)
+                {
+                    return;
+                }
+
+                context.RegisterSyntaxNodeAction(context => AnalyzeSyntax(context, argumentNullException, referenceEqualsMethod), SyntaxKind.MethodDeclaration);
+            });
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, ITypeSymbol argumentNullException, IMethodSymbol referenceEqualsMethod)
+        {
+            var cancellationToken = context.CancellationToken;
+
+            var semanticModel = context.SemanticModel!;
+            var syntaxTree = semanticModel.SyntaxTree;
```

Comment:
Just to avoid null suppression.

```suggestion
            var syntaxTree = context.Node.SyntaxTree;
```